apisix/secret/gcp.lua

```diff
@@ -0,0 +1,199 @@
+                entries_uri = {
+                    type = "string",
+                    default = "https://secretmanager.googleapis.com/v1"
+                },
+            },
+            required = { "client_email", "private_key", "project_id" }
+        },
+        ssl_verify = {
+            type = "boolean",
+            default = true
+        },
+        auth_file = { type = "string" },
+    },
+    oneOf = {
+        { required = { "auth_config" } },
+        { required = { "auth_file" } },
+    },
+}
+
+local _M = {
+    schema = schema
+}
+
+local function fetch_oauth_conf(conf)
+    if conf.auth_config then
+        return conf.auth_config
+    end
+
+    local file_content, err = core.io.get_file(conf.auth_file)
+    if not file_content then
+        return nil, "failed to read configuration, file: " .. conf.auth_file .. ", err: " .. err
+    end
+
+    local config_tab, err = core.json.decode(file_content)
+    if not config_tab then
+        return nil, "config parse failure, data: " .. file_content .. ", err: " .. err
+    end
+
+    local config = {
+        auth_config = {
+            client_email = config_tab.client_email,
+            private_key = config_tab.private_key,
+            project_id = config_tab.project_id
+        }
+    }
+
+    local ok, err = core.schema.check(schema, config)
+    if not ok then
+        return nil, "config parse failure, file: " .. conf.auth_file .. ", err: " .. err
+    end
+
+    return config_tab
+end
+
+
+local function get_secret(oauth, secrets_id)
+    local http_new = http.new()
+
+    local access_token = oauth:generate_access_token()
+    if not access_token then
+        return nil, "failed to get google oauth token"
+    end
+
+    local entries_uri = oauth.entries_uri .. "/projects/" .. oauth.project_id
+                            .. "/secrets/" .. secrets_id .. "/versions/latest:access"
+
+    local res, err = http_new:request_uri(entries_uri, {
+        ssl_verify = oauth.ssl_verify,
+        method = "GET",
+        headers = {
+            ["Content-Type"] = "application/json",
+            ["Authorization"] = (oauth.access_token_type or "Bearer") .. " " .. access_token,
+        },
+    })
+
+    if not res then
+        return nil, err
+    end
+
+    if res.status ~= 200 then
+        return nil, res.body
+    end
+
+    local body, err = core.json.decode(res.body)
+    if not body then
+        return nil, "failed to parse response data, " .. err
+    end
+
+    local payload = body.payload
+    if not payload then
+        return nil, "invalid payload"
+    end
+
+    return decode_base64(payload.data)
+end
+
+
+local function request_to_gcp(conf, secrets_id)
+    local auth_config, err = fetch_oauth_conf(conf)
+    if not auth_config then
+        return nil, err
+    end
+
+    local lru_key =  auth_config.client_email .. "#" .. auth_config.project_id
+
+    local oauth, err = lrucache(lru_key, "gcp", google_oauth.new, auth_config, conf.ssl_verify)
```

Comment:
the lrucache version will always be the same since "gcp" is a constant. Are you sure this is an intended behaviour?

Reply:
I need an LRU cache to cache my OAuth Token, which is stateful. Therefore, it should always be the same LRU cache, and the version of the LRU cache should be consistent.